@here/harp-mapview/lib/text/TextElementsRenderer.ts

```diff
@@ -1665,22 +1678,27 @@ export class TextElementsRenderer {
                 iconIndex === undefined
             );
             if (placeResult === PlacementResult.Invisible) {
-                if (placementStats) {
+                if (placementStats !== undefined) {
                     placementStats.numPoiTextsInvisible++;
                 }
-                labelState.reset();
-                return false;
+                if (!renderIcon || iconInvisible) {
+                    labelState.reset();
+                    return false;
+                }
             }
 
             const textRejected = placeResult === PlacementResult.Rejected;
-            if (!iconRejected) {
+            if (!iconRejected && !iconInvisible) {
                 const textIsOptional: boolean =
                     pointLabel.poiInfo !== undefined && pointLabel.poiInfo.textIsOptional === true;
                 iconRejected = textRejected && !textIsOptional;
             }
 
             if (textRejected) {
-                textRenderState!.startFadeOut(renderParams.time);
+                if (textRenderState!.isReallyVisible()) {
```

Comment:
This check shouldn't be needed.

If the label is FadingOut or FadedOut, startFadeOut() does nothing. We should add a check there to also return if the state is undefined.

Reply:
You forgot to add the check for Undefined state in startFadeOut()

Reply:
A label in Undefined state is not visible, it doesn't make sense for it to fade out, only to fade in.

Reply:
Modified startFadeOut() and the unit tests accordingly. I hope the rest of the code plays along.